concrete/src/Express/Association/Formatter/AbstractFormatter.php

```diff
@@ -33,10 +33,35 @@ public function getEntryDisplayName(AssociationControl $control, Entry $entry)
             } catch (\Exception $e) {
             }
         }
-        $targetEntity = $this->association->getTargetEntity();
-        $attribute = $targetEntity->getAttributes()[0];
+
+        $entity = $entry->getEntity();
+
+        /** @var \Concrete\Core\Entity\Attribute\Key\ExpressKey[] $attributes */
+        $attributes = $entity->getAttributes();
+
+        if ($keyId = $entity->getDisplayAttributeKey()) {
+
+            foreach ($attributes as $key) {
+                if ($key->getAttributeKeyID() == $keyId) {
+                    $attribute = $key;
+                    break;
+                }
+            }
+        } else {
+            foreach ($attributes as $attribute) {
+                if ($attribute->getAttributeType()->getAttributeTypeHandle() === 'text') {
+                    break;
+                }
+            }
+        }
+
         if (is_object($attribute)) {
-            return $entry->getAttribute($attribute);
+            $result = $entry->getAttribute($attribute);
+            if (!$result || (!is_string($result) && !is_callable($result, '__toString'))) {
+                return t('[Invalid Name]');
```

Comment:
Me too: I think it's the first time we use square brackets in the core in this way.

I think it's fine to take them out the `t()` argument: in this case they are not part of a sentence, they are just a graphical decoration